pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -805,25 +929,50 @@ func (n *NetworkPolicyController) addService(obj interface{}) {
 	}
+
+	// If this Service doesn't have a selector, there won't be an Endpoints ADD event
+	// triggered by this Service ADD event. So find all ACNPs that are impacted by this
+	// Service and reprocess them here. Otherwise, an Endpoints ADD event will
+	// reprocess those ACNPs.
+	if service.Spec.Selector == nil {
+		affectedACNPsByService := n.filterACNPsByService(service.Namespace, service.Name)
+		for _, acnp := range affectedACNPsByService {
+			n.reprocessACNP(acnp)
+		}
+	}
 }
 
-// updatePod retrieves all internal Groups which refers to this Service
-// and enqueues the group keys for further processing.
+// updateService retrieves all internal Groups which refers to this Service
+// and enqueues the group keys for further processing. And reprocess the
+// ClusterNetworkPolicy with `toService` referred to this Service.
 func (n *NetworkPolicyController) updateService(oldObj, curObj interface{}) {
 	defer n.heartbeat("updateService")
 	oldService := oldObj.(*v1.Service)
 	curService := curObj.(*v1.Service)
 	klog.V(2).Infof("Processing Service %s/%s UPDATE event, selectors: %v", curService.Namespace, curService.Name, curService.Spec.Selector)
-	// No need to trigger processing of groups if there is no change in the Service selectors.
-	if reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
-		klog.V(4).Infof("No change in Service %s/%s. Skipping group evaluation.", curService.Namespace, curService.Name)
+	// Only trigger processing of groups if there are changes in the Service selectors.
+	if !reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
+		// Find all internal Group keys which refers to this Service.
+		groupKeySet := n.filterInternalGroupsForService(curService)
+		// Enqueue internal groups to its queue for group processing.
+		for group := range groupKeySet {
+			n.enqueueInternalGroup(group)
+		}
+		// If the Selector of this Service has been updated, an Endpoints UPDATE event will
+		// reprocess all ACNPs that are impacted by this Service. No need to reprocess them
+		// here.
 		return
 	}
-	// Find all internal Group keys which refers to this Service.
-	groupKeySet := n.filterInternalGroupsForService(curService)
-	// Enqueue internal groups to its queue for group processing.
-	for group := range groupKeySet {
-		n.enqueueInternalGroup(group)
+
+	// If this Service doesn't have a Selector or both Selector and Ports haven't been
+	// updated, then there won't be an Endpoints UPDATE event triggered by this Service
+	// UPDATE event. So find all ACNPs that are impacted by this Service and reprocess
+	// them here. Otherwise, an Endpoints UPDATE event will reprocess those ACNPs.
+	if curService.Spec.Selector == nil || (curService.Spec.Selector != nil && reflect.DeepEqual(oldService.Spec.Ports, curService.Spec.Ports)) {
+		affectedACNPsByService := n.filterACNPsByService(curService.Namespace, curService.Name)
+		for _, acnp := range affectedACNPsByService {
+			n.reprocessACNP(acnp)
+		}
```

Comment:
this is how I view things: all the Policy rules that use the same `toService` reference will end up (or at least should end up) using the same AddressGroup. When a Service selector is updated, it seems that I just need to update the contents of this AddressGroup as needed. There is no need for an Index or for the `filterACNPsByService` IMO, and we don't need to enqueue potentially a large number of ACNPs. This should really be more similar to ClusterGroups I think.

Adding @tnqn as he will know better than me.

Reply:
I didn't give it as much thought but I feel like the same applies to NodePort Services. We just need 2 identifiable AddressGroups in this case: one with Endpoint IPs and one with Node IPs.

Reply:
I realized that this doesn't cover the case where the target port is updated, so the index may be necessary. Still believe we should handle NodePort in a separate PR.

Reply:
NodePort case removed.